File: packages/wasmvm/wasmlib/src/wasmtypes/scstring.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::convert::TryInto;

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn string_decode(dec: &mut WasmDecoder) -> String {
    let length = uint16_decode(dec);
    string_from_bytes(&dec.fixed_bytes(length as usize))
}

pub fn string_encode(enc: &mut WasmEncoder, value: &str) {
    let buf = string_to_bytes(value);
    uint16_encode(enc, buf.len() as u16);
    enc.fixed_bytes(&buf, buf.len());
}

pub fn string_from_bytes(buf: &[u8]) -> String {
    unsafe {
        String::from_utf8_unchecked(buf.to_vec()).to_string()
    }
}

pub fn string_to_bytes(value: &str) -> Vec<u8> {
    value.as_bytes().to_vec()
}

pub fn string_from_string(value: &str) -> String {
    value.to_string()
}

pub fn string_to_string(value: &str) -> String {
    value.to_string()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableString {
    proxy: Proxy,
}

impl ScImmutableString {
    pub fn new(proxy: Proxy) -> ScImmutableString {
        ScImmutableString { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        self.value()
    }

    pub fn value(&self) -> String {
        string_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable ScString in host container
pub struct ScMutableString {
    proxy: Proxy,
}

impl ScMutableString {
    pub fn new(proxy: Proxy) -> ScMutableString {
        ScMutableString { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: &str) {
        self.proxy.set(&string_to_bytes(value));
    }

    pub fn to_string(&self) -> String {
        self.value()
    }

    pub fn value(&self) -> String {
        string_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/scstring.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';
import {bytesFromUint8Array} from './scbytes';
import {uint16Decode, uint16Encode} from "./scuint16";

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function stringDecode(dec: WasmDecoder): string {
    const length = uint16Decode(dec);
    return stringFromBytes(dec.fixedBytes(length as u32));
}

export function stringEncode(enc: WasmEncoder, value: string): void {
    const buf = stringToBytes(value);
    uint16Encode(enc, buf.length as u16);
    enc.fixedBytes(buf, buf.length as u32);
}

export function stringFromBytes(buf: Uint8Array): string {
    return String.UTF8.decodeUnsafe(buf.dataStart, buf.length);
}

export function stringToBytes(value: string): Uint8Array {
    const arrayBuffer = String.UTF8.encode(value);
    const u8Array = Uint8Array.wrap(arrayBuffer);
    return bytesFromUint8Array(u8Array);
}

export function stringFromString(value: string): string {
    return value;
}

export function stringToString(value: string): string {
    return value;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableString {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return this.value();
    }

    value(): string {
        return stringFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableString extends ScImmutableString {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: string): void {
        this.proxy.set(stringToBytes(value));
    }
}
